Conservative transfers

fdkong opened the pull request:
This is a simple implementation of conservative transfers based on my understanding. 

But I do not think I completely understand the requirements described in #12948 by @vincentlaboure. I am not going to add more code before I understand more since I may be going in a wrong direction. 

@YaqiWang may be also interested. 


<!--
INCLUDE THE FOLLOWING IN THE PR DESCRIPTION
- Explain relevant design information for your change.
- Follow the [Coding Standards](http://mooseframework.org/wiki/CodeStandards/).
- Submit or improve [Test Cases](http://mooseframework.org/wiki/MooseTraining/testing/).
- Reference a specific issue, place "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).
-->

fdkong committed:
The first simple implementation of a conservative transfer

Refs #12948

fdkong committed:
Conservative transfers: added a test

that uses MultiAppCopyTransfer. But the code should work for other field transferes

Refs #12948

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/304696/) on 688ffcd : invalidated by @permcody

friedmud commented:
I would definitely like a chance to review this before it goes in...

On Wed, Mar 13, 2019 at 6:56 PM MOOSE Build <notifications@github.com>
wrote:

> Job Precheck <https://civet.inl.gov/job/304696/> on 688ffcd
> <https://github.com/idaholab/moose/commit/688ffcd3d48afdc823c8a6550ffa1096b0470ac7>
> : invalidated by @permcody <https://github.com/permcody>
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/13060#issuecomment-472651587>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AA1JMW37KwWf1ihDCvHf1yH83L29CF5eks5vWZApgaJpZM4bxzTE>
> .
>
-- 
Sent from my iPhone

fdkong commented:
The implementation is not done yet. I posted the code here to have @vincentlaboure describe his requirements in detail.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/304702/) on 688ffcd wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/13060/site)

This comment will be updated on new commits.

fdkong committed:
MultiAppTransfer: Fixed unity compiling errors

fdkong committed:
Excute one single user object

Refs #12948

fdkong committed:
Added a field transfer interface so that

the caculations can be reused for several interpolation transfers

Refs #12948

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/307466/) on f8f1a6f wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/13060/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/13060/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 2dc9b9605b1a8a437395eaf57875ebb7df78cebf`

permcody committed:
Get rid of the run-away RegEx

YaqiWang commented on framework/src/transfers/MultiAppFieldTransferInterface.C:
```diff
@@ -0,0 +1,102 @@
+  InputParameters params = validParams<MultiAppTransfer>();
+  params.addRequiredParam<AuxVariableName>(
+      "variable", "The auxiliary variable to store the transferred values in.");
+  params.addRequiredParam<VariableName>("source_variable", "The variable to transfer from.");
```
Is it possible to make this parameter and the above as vector of variable names?

fdkong replied:
Definitely, it is possible. But from my understanding, most current field transfers such as the interpolation transfer and the nearestnode  transfer take one nonlinear variable and transfer it to an auxiliary. My implementation simply follow this style.  

If you want to transfer  multiple variables, and you could just simply create multiple transfers in  your input file.

YaqiWang commented on framework/src/transfers/MultiAppFieldTransferInterface.C:
```diff
@@ -0,0 +1,102 @@
+validParams<MultiAppFieldTransferInterface>()
+{
+  InputParameters params = validParams<MultiAppTransfer>();
+  params.addRequiredParam<AuxVariableName>(
```
Is the restriction on aux var necessary? Looks like the code can work with nonlinear variables.

fdkong replied:
Transfers could happen in anytime (nonlinear, linear). The solution, in theory, is locked by PETSc, and we should not change that vector anyway (you definitely  can if you really want). 

The solution vector in the auxiliary system is accessible since it is not touched by solvers.

I could be totally wrong on this, and there are other reasons @friedmud  and @permcody may be aware of.

vincentlaboure committed:
Add examples with desired conservative transfer

fdkong committed:
Make to_mesh and from_mesh the same since

MultiAppNearestNodeTransfer produces different results with
different number of processor cores.

The fix for MultiAppNearestNodeTransfer is challenging due to geometric search

fdkong committed:
Restrict ``transfer" to postprocessor

fdkong committed:
Added one example using MultiAppMeshFunctionTransfer

with conservative feature

fdkong committed:
Changed gold files

Refs #12948

fdkong committed:
Support one variable only for the conservative transfer

Refs #12948

fdkong committed:
Added NearestPointIntegralVariablePostprocessor

The pps works for the conservative transfer now

Refs #12948

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/325851/) on 38f77bd wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/13060/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/13060/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 0595a2b2744d71d60aa8cbb736a908656a6ad69f`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/326901/) on 11b1667 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/13060/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/13060/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 0595a2b2744d71d60aa8cbb736a908656a6ad69f`

fdkong committed:
Make the base class of NearestPointBase as a template

Refs #13252

fdkong committed:
Added NearestPointLayeredSideAverage

It works now

Refs #13252

fdkong committed:
Fixed CHPFCRFFSplitVariablesAction since we had

made the transfer to support multible variables

Refs #13252

fdkong committed:
Added docs for NearestPointLayeredSideAverage

fdkong committed:
gatherSum(_integral_value) should not be in getValue()

otherwise, you will get a different value when call
getValue() again

Refs #13252

fdkong commented:
@vincentlaboure I have implemented the conservative transfer and NearestPointLayeredSideAverage.

Could you take a look at? If these functions are exactly the ones you need, I will fix the documentation and have a merge process. Otherwise, please let us what we are missing here?

YaqiWang commented:
`NearestPointLayeredSideAverage` could be something needed by @snschune.

snschune commented:
> NearestPointLayeredSideAverage could be something needed by

Yes that is indeed important!

fdkong commented:
> Yes that is indeed important!

I already implemented it. Could  you take a look if it is what you need.

snschune commented:
@fdkong That looks like what I need. I'll test it once it goes in.

vincentlaboure commented:
It will be very useful, thanks @fdkong for working on it!

I am going to try the test problems you have and see if I have comments

fdkong commented:
> It will be very useful, thanks @fdkong for working on it!
> 
> I am going to try the test problems you have and see if I have comments

Thanks.

vincentlaboure commented on framework/src/transfers/MultiAppFieldTransferInterface.C:
```diff
@@ -0,0 +1,396 @@
+          mooseError(
+              " You have to specify ",
+              _multi_app->numGlobalApps(),
+              " regular from-postproessors, or use NearestPointIntegralVariablePostprocessor ");
```
typo: `from-postproessors` -> `from-postprocessors`

vincentlaboure commented on framework/src/transfers/MultiAppFieldTransferInterface.C:
```diff
@@ -0,0 +1,396 @@
+          mooseError(
+              " You have to specify ",
+              _multi_app->numGlobalApps(),
+              " regular to-postproessors, or use NearestPointIntegralVariablePostprocessor ");
```
same

vincentlaboure commented on framework/src/transfers/MultiAppFieldTransferInterface.C:
```diff
@@ -0,0 +1,396 @@
+  }
+
+  PostprocessorValue to_adjuster = 0;
+  // Compute to-postproessor to have the adjuster
```
`to-postproessor` -> `to-postprocessor`

YaqiWang commented on framework/src/postprocessors/ElementIntegralPostprocessor.C:
```diff
@@ -39,10 +39,15 @@ ElementIntegralPostprocessor::execute()
 }
 
+void
+ElementIntegralPostprocessor::finalize()
```
👍 This is something I wanted for a long time.

YaqiWang commented on framework/src/transfers/MultiAppDTKInterpolationTransfer.C:
```diff
@@ -24,19 +24,16 @@ template <>
-    _to_var_name(getParam<AuxVariableName>("variable"))
+  : MultiAppFieldTransferInterface(parameters)
 {
+  mooseAssert(_to_var_name.size() == 1 && _from_var_name.size() == 1,
```
Better use mooseError here I think.

YaqiWang commented on framework/src/transfers/MultiAppFieldTransferInterface.C:
```diff
@@ -0,0 +1,396 @@
+  return params;
+}
+
+MultiAppFieldTransferInterface::MultiAppFieldTransferInterface(const InputParameters & parameters)
```
Should we just name this as `MultiAppFieldTransfer` since it is in the inheritance chain but not a separate interface?

YaqiWang commented on framework/src/transfers/MultiAppFieldTransferInterface.C:
```diff
@@ -0,0 +1,396 @@
+     * Not sure how important to support multi variables
+     * Let us handle the single variable case only right now if the conservative capability is on
+     */
+    mooseAssert(_to_var_name.size() == 1 && _from_var_name.size() == 1,
```
mooseError is preferred.

YaqiWang replied:
Will not comment on each individual mooseAssert. I think most of them should be mooseError.

vincentlaboure commented on test/tests/transfers/multiapp_conservative_transfer/master_nearest_point.i:
```diff
@@ -0,0 +1,134 @@
+
+[Postprocessors]
+  [from_nearest_point]
+    type = NearestPointIntegralVariablePostprocessor
```
It looks like the value of the `NearestPointIntegralVariablePostprocessor` pps are not printed to screen (or in the CSV file with `csv=true`). It would be useful to have it, especially to check that the pps indeed match.

vincentlaboure replied:
This still seems to be a problem

fdkong replied:
I do not get it here. What do you mean by "still a problem"? Could you tell what you want?

fdkong replied:
NearestPointIntegralVariablePostprocessor now is a standard vector post processor, and its values can be printed to a `csv` file if you you have `"csv=true"`

vincentlaboure replied:
For some reason, the value of this postprocessor never appears to my screen:

```
Running App: main
Framework Information:
MOOSE Version:           git commit 8f529ee8dc on 2019-05-17
LibMesh Version:         c4c9fd5450489486fd5f7baf2ec9a8ac0d47fc99
PETSc Version:           3.9.4
Current Time:            Fri May 17 12:10:12 2019
Executable Timestamp:    Fri May 17 11:46:44 2019

Parallelism:
  Num Processors:          1
  Num Threads:             1

Mesh: 
  Parallel Type:           replicated
  Mesh Dimension:          2
  Spatial Dimension:       2
  Nodes:                   
    Total:                 121
    Local:                 121
  Elems:                   
    Total:                 100
    Local:                 100
  Num Subdomains:          2
  Num Partitions:          1

Nonlinear System:
  Num DOFs:                121
  Num Local DOFs:          121
  Variables:               "power_density" 
  Finite Element Types:    "LAGRANGE" 
  Approximation Orders:    "FIRST" 

Auxiliary System:
  Num DOFs:                121
  Num Local DOFs:          121
  Variables:               "from_sub" 
  Finite Element Types:    "LAGRANGE" 
  Approximation Orders:    "FIRST" 

Execution Information:
  Executioner:             Steady
  Solver Mode:             Preconditioned JFNK
  Preconditioner:          hypre boomeramg 

Initializing All MultiApps
Creating MultiApp sub of type MooseTestApp of level 1 and number 0:
sub0: 
Creating MultiApp sub of type MooseTestApp of level 1 and number 1:
sub1: 
sub0: 
sub0: Running App: sub0
sub0: Parallelism:
sub0:   Num Processors:          1
sub0:   Num Threads:             1
sub0: 
sub0: Mesh: 
sub0:   Parallel Type:           replicated
sub0:   Mesh Dimension:          2
sub0:   Spatial Dimension:       2
sub0:   Nodes:                   
sub0:     Total:                 121
sub0:     Local:                 121
sub0:   Elems:                   
sub0:     Total:                 100
sub0:     Local:                 100
sub0:   Num Subdomains:          2
sub0:   Num Partitions:          1
sub0: 
sub0: Nonlinear System:
sub0:   Num DOFs:                100
sub0:   Num Local DOFs:          100
sub0:   Variables:               "sink" 
sub0:   Finite Element Types:    "MONOMIAL" 
sub0:   Approximation Orders:    "CONSTANT" 
sub0: 
sub0: Auxiliary System:
sub0:   Num DOFs:                21
sub0:   Num Local DOFs:          21
sub0:   Variables:               "from_master" 
sub0:   Finite Element Types:    "LAGRANGE" 
sub0:   Approximation Orders:    "FIRST" 
sub0: 
sub0: Execution Information:
sub0:   Executioner:             Steady
sub0:   Solver Mode:             Preconditioned JFNK
sub0:   Preconditioner:          hypre boomeramg 
sub0: 
sub1: 
sub1: Running App: sub1
sub1: Parallelism:
sub1:   Num Processors:          1
sub1:   Num Threads:             1
sub1: 
sub1: Mesh: 
sub1:   Parallel Type:           replicated
sub1:   Mesh Dimension:          2
sub1:   Spatial Dimension:       2
sub1:   Nodes:                   
sub1:     Total:                 121
sub1:     Local:                 121
sub1:   Elems:                   
sub1:     Total:                 100
sub1:     Local:                 100
sub1:   Num Subdomains:          2
sub1:   Num Partitions:          1
sub1: 
sub1: Nonlinear System:
sub1:   Num DOFs:                100
sub1:   Num Local DOFs:          100
sub1:   Variables:               "sink" 
sub1:   Finite Element Types:    "MONOMIAL" 
sub1:   Approximation Orders:    "CONSTANT" 
sub1: 
sub1: Auxiliary System:
sub1:   Num DOFs:                21
sub1:   Num Local DOFs:          21
sub1:   Variables:               "from_master" 
sub1:   Finite Element Types:    "LAGRANGE" 
sub1:   Approximation Orders:    "FIRST" 
sub1: 
sub1: Execution Information:
sub1:   Executioner:             Steady
sub1:   Solver Mode:             Preconditioned JFNK
sub1:   Preconditioner:          hypre boomeramg 
sub1: 
Finished Initializing All MultiApps

Backing Up MultiApps on TIMESTEP_END
Beginning backing up MultiApp sub
Finished backing up MultiApp sub
Finished Backing Up MultiApps on TIMESTEP_END

 0 Nonlinear |R| = 3.317228e+03
      0 Linear |R| = 3.317228e+03
      1 Linear |R| = 8.653417e+01
      2 Linear |R| = 2.749651e+00
      3 Linear |R| = 8.942078e-02
      4 Linear |R| = 2.238424e-03
 1 Nonlinear |R| = 6.375386e-02
      0 Linear |R| = 6.375386e-02
      1 Linear |R| = 1.596349e-03
      2 Linear |R| = 4.511555e-05
      3 Linear |R| = 1.604996e-06
      4 Linear |R| = 4.238846e-08
 2 Nonlinear |R| = 1.147928e-07
 Solve Converged!

Starting Transfers on TIMESTEP_END To MultiApps
Beginning MeshFunctionTransfer to_sub
Finished MeshFunctionTransfer to_sub
Beginning Conservative transfers to_sub
Finished Conservative transfers to_sub
Transfers on TIMESTEP_END Are Finished


Executing MultiApps on TIMESTEP_END
sub0: 
sub0: Postprocessor Values:
sub0: +----------------+----------------+----------------+
sub0: | time           | from_master_pp | sink           |
sub0: +----------------+----------------+----------------+
sub0: |   0.000000e+00 |   1.811928e+01 |   0.000000e+00 |
sub0: +----------------+----------------+----------------+
sub0: 
sub0:  0 Nonlinear |R| = 3.467168e+00
sub0:       0 Linear |R| = 3.467168e+00
sub0:       1 Linear |R| = 9.331840e-06
sub0:  1 Nonlinear |R| = 2.712319e-06
sub0:       0 Linear |R| = 2.712319e-06
sub0:       1 Linear |R| = 3.594458e-13
sub0:  2 Nonlinear |R| = 3.842817e-13
sub0:  Solve Converged!
sub0: 
sub0: Postprocessor Values:
sub0: +----------------+----------------+----------------+
sub0: | time           | from_master_pp | sink           |
sub0: +----------------+----------------+----------------+
sub0: |   0.000000e+00 |   1.811928e+01 |   0.000000e+00 |
sub0: |   1.000000e+00 |   1.538888e+02 |   4.663411e+00 |
sub0: +----------------+----------------+----------------+
sub0: 
sub1: 
sub1: Postprocessor Values:
sub1: +----------------+----------------+----------------+
sub1: | time           | from_master_pp | sink           |
sub1: +----------------+----------------+----------------+
sub1: |   0.000000e+00 |   4.691928e+01 |   0.000000e+00 |
sub1: +----------------+----------------+----------------+
sub1: 
sub1:  0 Nonlinear |R| = 3.467168e+00
sub1:       0 Linear |R| = 3.467168e+00
sub1:       1 Linear |R| = 9.331840e-06
sub1:  1 Nonlinear |R| = 2.712319e-06
sub1:       0 Linear |R| = 2.712319e-06
sub1:       1 Linear |R| = 3.594458e-13
sub1:  2 Nonlinear |R| = 3.842817e-13
sub1:  Solve Converged!
sub1: 
sub1: Postprocessor Values:
sub1: +----------------+----------------+----------------+
sub1: | time           | from_master_pp | sink           |
sub1: +----------------+----------------+----------------+
sub1: |   0.000000e+00 |   4.691928e+01 |   0.000000e+00 |
sub1: |   1.000000e+00 |   4.038887e+02 |   4.663411e+00 |
sub1: +----------------+----------------+----------------+
sub1: 
Finished Executing MultiApps on TIMESTEP_END


Starting Transfers on TIMESTEP_END From MultiApps
Beginning MeshFunctionTransfer from_sub
Finished MeshFunctionTransfer from_sub
Beginning Conservative transfers from_sub
Finished Conservative transfers from_sub
Transfers on TIMESTEP_END Are Finished


sub1: 
sub0: 
```

Is that normal?

vincentlaboure replied:
It makes it hard to check that the transfers are indeed conservative

YaqiWang commented on framework/src/transfers/MultiAppFieldTransferInterface.C:
```diff
@@ -0,0 +1,396 @@
+    }
+  }
+
+  mooseAssert(_to_var_name.size() == _from_var_name.size() && _to_var_name.size() >= 1,
```
`_to_var_name.size() >= 1` looks not necessary.

vincentlaboure reviewed (commented):
Some comments after trying the tests in `test/tests/transfers/multiapp_conservative_transfer`

Thanks!

YaqiWang commented on framework/src/transfers/MultiAppFieldTransferInterface.C:
```diff
@@ -0,0 +1,396 @@
+{
+  InputParameters params = validParams<MultiAppTransfer>();
+  params.addRequiredParam<std::vector<AuxVariableName>>(
+      "variable", "The auxiliary variable to store the transferred values in.");
```
This is a quite big thing. Should we allow the target variable to be a primal variable? I think we should.

YaqiWang commented on framework/src/transfers/MultiAppFieldTransferInterface.C:
```diff
@@ -0,0 +1,396 @@
+      "variable", "The auxiliary variable to store the transferred values in.");
+  params.addRequiredParam<std::vector<VariableName>>("source_variable",
+                                                     "The variable to transfer from.");
+  params.addParam<bool>("preserve_transfer",
```
I do not quite like the user interface here and have a suggestion. We only have `from_postprocessor_to_be_preserved` and `to_postprocessor_to_be_preserved` but make them optional without default values. Then we do preserve treatment only when they are valid. We do not need this bool parameter.

YaqiWang commented on framework/src/transfers/MultiAppFieldTransferInterface.C:
```diff
@@ -0,0 +1,396 @@
+  : MultiAppTransfer(parameters),
+    _from_var_name(getParam<std::vector<VariableName>>("source_variable")),
+    _to_var_name(getParam<std::vector<AuxVariableName>>("variable")),
+    _preserve_transfer(parameters.get<bool>("preserve_transfer")),
```
Then here we can just have `_preserve_transfer(isParamValid("from_postprocessor_to_be_preserved"")),`.

fdkong replied:
OK

YaqiWang commented on framework/src/transfers/MultiAppFieldTransferInterface.C:
```diff
@@ -0,0 +1,396 @@
+    if (_from_postprocessor_to_be_preserved.size() == 1 && _direction == TO_MULTIAPP)
+    {
+      FEProblemBase & from_problem = _multi_app->problemBase();
+      auto * pps = dynamic_cast<const NearestPointIntegralVariablePostprocessor *>(
```
Remember we decided to have VectorPostprocessor but why here we are using `NearestPointIntegralVariablePostprocessor`?

fdkong replied:
Really? I remembered last time we decided to use `NearestPointIntegralVariablePostprocessor`.  I could add one more support if it is necessary.

YaqiWang replied:
Thought that pp is for Sebastian's model only. @friedmud mentioned few functions in the UserObject base class can be useful. I might be wrong. But I feel strange to let nearestpoint stuff control transfer preserving here. VectorPostprocessor is nice to user here but not a necessity.

YaqiWang commented on framework/src/transfers/MultiAppInterpolationTransfer.C:
```diff
@@ -52,25 +48,28 @@ validParams<MultiAppInterpolationTransfer>()
+
   if (_direction == TO_MULTIAPP)
-    variableIntegrityCheck(_to_var_name);
+    variableIntegrityCheck(_to_var_name[0]);
```
this integrity check looks to me should be in `MultiAppFieldTransferInterface::initialSetup`. Then here we do not need to override the functionl

fdkong replied:
Right

YaqiWang commented on framework/src/transfers/MultiAppNearestNodeTransfer.C:
```diff
@@ -411,7 +413,7 @@ MultiAppNearestNodeTransfer::execute()
         MooseVariableFEBase & from_var = _from_problems[_cached_froms[i_proc][qp]]->getVariable(
             0,
-            _from_var_name,
+            _from_var_name[0],
```
This is quite annoying. We can create `_from_var_name` in the base class and rename the vector one as `_from_var_names`. The `_from_var_name` is equal to `_from_var_names[0]`. Another suggestion is to create a call back function in the base class like `bool allowMultipleVariables() {return true; }` and let those transfers do not support multiple variables `return false`.

fdkong replied:
Good point.

YaqiWang commented on framework/src/transfers/MultiAppTransfer.C:
```diff
@@ -12,6 +12,7 @@
 #include "Transfer.h"
 #include "MooseTypes.h"
 #include "FEProblem.h"
+#include "NonlinearSystemBase.h"
```
needed?

YaqiWang commented:
@fdkong I am done reviewing. I did not check the nearestnode related stuff. This is good. My only big concern is the variable type of the targeting variable. I think we should allow it to be primal variable as well.

fdkong commented on test/tests/transfers/multiapp_conservative_transfer/master_power_density.i:
```diff
@@ -0,0 +1,143 @@
+    from_postprocessor_to_be_preserved = 'sink'
+  []
+[]
+
```
I knew this thing, but I have no idea to make it look right. I will try again. It is related to console

fdkong commented:
> @fdkong I am done reviewing. I did not check the nearestnode related stuff. This is good. My only big concern is the variable type of the targeting variable. I think we should allow it to be primal variable as well.

Just talked to @permcody and @friedmud. We need a user case to demonstrate why we want to transfer from one nonlinear variable to another nonlinear variable.  The code is actually very easy to support this.

My concern would be: we can not change the solution at some stages of the simulations, for example, we should not change the solution during the nonlinear functions evaluation and the Jacobian calculations.

YaqiWang commented:
My only use case in my mind is that let multiapp run on initial and transfer the variables to master primal variables as the initial condition or initial guess. I agree it is rare to transfer to primal variables.

fdkong commented:
> My only use case in my mind is that let multiapp run on initial and transfer the variables to master primal variables as the initial condition or initial guess. I agree it is rare to transfer to primal variables.

The initial guess can be passed in through the aux variable as well.  I would like to avoid changing the current design if this is only user case.

fdkong committed:
We compute to-processor again so that

its value is right with updated solution

Refs #12948

fdkong committed:
Conservative transfer: make the subapp mesh mismatch

the master mesh so that the test is more meaningful

Refs #12948

fdkong committed:
Conversative transfer: to use mooseError instead of mooseAssert

Refs #12948

fdkong committed:
Conservative transfer: clean up user interfaces

Refs #12948

fdkong committed:
Changed user interface to use from_postprocessors_to_be_preserved

instead of from_postprocessor_to_be_preserved

Refs #12948

fdkong committed:
Rename MultiAppFieldTransfer to MultiAppFieldTransferInterface

Refs #12948

fdkong committed:
Conservative transfers: be able to output adjusting values

Refs #12948

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/335298/) on 5659a52 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/13060/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/13060/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 0595a2b2744d71d60aa8cbb736a908656a6ad69f`

YaqiWang commented:
@fdkong are you still working on this PR?

fdkong commented:
> @fdkong are you still working on this PR?

Yes, I am adding more tests.

fdkong commented on framework/src/transfers/MultiAppFieldTransfer.C:
```diff
@@ -0,0 +1,417 @@
+     * Let us handle the single variable case only right now if the conservative capability is on
+     */
+    if (_to_var_name.size() != 1 && _from_var_names.size() != 1)
+      mooseError(" Support single variable only when the conservative capability is on ");
```
OK...

fdkong commented on framework/src/userobject/NearestPointIntegralVariablePostprocessor.C:
```diff
@@ -0,0 +1,85 @@
+std::shared_ptr<ElementIntegralVariablePostprocessor>
+NearestPointIntegralVariablePostprocessor::userObject(unsigned int i) const
+{
+  return _user_objects[i];
```
user_object was originally created in NearPointBase, and stored as a share pointer. I do not know why it was originally stored as a shared pointer. So I simply return this pointer. 

I could deference, and return a object if you like so.

fdkong commented on framework/src/transfers/MultiAppFieldTransfer.C:
```diff
@@ -0,0 +1,417 @@
+        auto ii = pps.nearestPointIndex(*node);
+        if (ii != i)
+          continue;
+        scale = from_adjuster / pps.userObjectValue(i);
```
It is possible. I will add checker.

fdkong committed:
Conservative transfer: addressed comments from Cody

fdkong commented:
I have no idea why docs is falling. Does  not look my changes are related to that.

@permcody

permcody commented on test/tests/transfers/multiapp_conservative_transfer/tests:
```diff
@@ -0,0 +1,33 @@
+    type = Exodiff
+    input = master_power_density.i
+    exodiff = 'master_power_density_out.e master_power_density_out_sub0.e master_power_density_out_sub1.e'
+    requirement = 'MOOSE shall support conservative transfers'
```
Requirements should be unique. You have two that are the same. They can be 90% the same, that's fine, what are you testing here and in the test above?

permcody commented on test/tests/transfers/multiapp_conservative_transfer/tests:
```diff
@@ -0,0 +1,33 @@
+    input = 'master_nearest_point.i'
+    csvdiff = 'master_nearest_point_out_from_nearest_point_0001.csv master_nearest_point_out_to_nearest_point_0001.csv'
+    prereq = conservative_transfer_nearest_point
+    requirement = 'MOOSE shall be able to output the adjusting values in conservative transfers'
```
Also, don't say MOOSE instead be more formal. "The system shall..."

permcody commented on framework/src/transfers/MultiAppFieldTransfer.C:
```diff
@@ -0,0 +1,443 @@
+        auto ii = pps.nearestPointIndex(*node);
+        if (ii != i)
+          continue;
+        if (pps.userObjectValue(i) == 0.0)
```
Equality checks with Real values aren't the best. Could we do a really small abs check with a tolerance. Or perhaps use one of the MOOSE "fuzzy" checks in the utils?

permcody commented:
This looks good - Let's fix up those requirements and I suggest that we do a better comparison against zero.

permcody commented:
FYI - @fdkong this is breaking the Trilinos build.

vincentlaboure commented:
Thanks @fdkong for the hard work on this!

permcody commented:
FYI - @fdkong this also appears to be breaking the new parallel tensor_mechanics tests (several) and the parallel xfem tests (several). They were passing on the previous merge.
